File: core-ditbrowse/protocol.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	log "github.com/s00500/env_logger"
)

const (
	controlProtocol        = "ditbrowse.control"
	controlProtocolVersion = 1
	controlWebSocketPath   = "/api/ws"
)

type ditBrowseStatusTab struct {
	Index        int     `json:"index"`
	TileID       string  `json:"tileId"`
	CameraID     *string `json:"cameraId"`
	CameraNumber *int    `json:"cameraNumber"`
	Title        string  `json:"title"`
	URL          string  `json:"url"`
}

type ditBrowseStatus struct {
	ExpansionEnabled     bool                 `json:"expansionEnabled"`
	FocusMode            bool                 `json:"focusMode"`
	SelectedCameraNumber *int                 `json:"selectedCameraNumber"`
	SelectedTileID       *string              `json:"selectedTileId"`
	SelectedIndex        *int                 `json:"selectedIndex"`
	Tabs                 []ditBrowseStatusTab `json:"tabs"`
}

type ditBrowseClient struct {
	host string
	port uint16

	mu     sync.Mutex
	conn   *websocket.Conn
	status ditBrowseStatus
	rev    int64

	onStatus func(ditBrowseStatus)
}

func newDitBrowseClient(host string, port uint16) *ditBrowseClient {
	return &ditBrowseClient{host: host, port: port}
}

func (c *ditBrowseClient) wsURL() string {
	return fmt.Sprintf("ws://%s:%d%s", c.host, c.port, controlWebSocketPath)
}

func (c *ditBrowseClient) connect() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn != nil {
		_ = c.conn.Close()
		c.conn = nil
	}

	dialer := websocket.Dialer{
		Proxy:            http.ProxyFromEnvironment,
		HandshakeTimeout: 5 * time.Second,
	}
	conn, _, err := dialer.Dial(c.wsURL(), nil)
	if err != nil {
		return err
	}

	hello := map[string]any{
		"type":            "hello",
		"protocol":        controlProtocol,
		"protocolVersion": controlProtocolVersion,
		"client": map[string]string{
			"name":    "core-ditbrowse",
			"version": gitTag,
		},
	}
	if err := conn.WriteJSON(hello); err != nil {
		_ = conn.Close()
		return err
	}

	_ = conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	var response map[string]any
	if err := conn.ReadJSON(&response); err != nil {
		_ = conn.Close()
		return fmt.Errorf("hello response: %w", err)
	}
	_ = conn.SetReadDeadline(time.Time{})

	if response["type"] != "hello" || response["protocol"] != controlProtocol {
		_ = conn.Close()
		return fmt.Errorf("unexpected hello response: %v", response["type"])
	}

	c.conn = conn
	return nil
}

func (c *ditBrowseClient) close() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn != nil {
		_ = c.conn.Close()
		c.conn = nil
	}
}

func (c *ditBrowseClient) readLoop(onDisconnect func()) {
	for {
		c.mu.Lock()
		conn := c.conn
		c.mu.Unlock()
		if conn == nil {
			onDisconnect()
			return
		}

		_, data, err := conn.ReadMessage()
		if err != nil {
			log.Warnf("DIT Browse read error from %s: %v", c.wsURL(), err)
			c.close()
			onDisconnect()
			return
		}

		var envelope struct {
			Type     string           `json:"type"`
			Event    string           `json:"event"`
			Revision int64            `json:"revision"`
			Status   *ditBrowseStatus `json:"status"`
			Ok       *bool            `json:"ok"`
		}
		if err := json.Unmarshal(data, &envelope); err != nil {
			log.Warnf("DIT Browse invalid JSON: %v", err)
			continue
		}

		if envelope.Type == "event" && envelope.Event == "status" && envelope.Status != nil {
			c.applyStatus(*envelope.Status, envelope.Revision)
			continue
		}

		if envelope.Type == "result" && envelope.Ok != nil && *envelope.Ok && envelope.Status != nil {
			c.applyStatus(*envelope.Status, c.rev)
		}
	}
}

func (c *ditBrowseClient) applyStatus(status ditBrowseStatus, revision int64) {
	c.mu.Lock()
	c.status = status
	if revision > c.rev {
		c.rev = revision
	}
	cb := c.onStatus
	c.mu.Unlock()
	if cb != nil {
		cb(status)
	}
}

func (c *ditBrowseClient) getStatus() ditBrowseStatus {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.status
}

func (c *ditBrowseClient) sendCommand(command map[string]any) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn == nil {
		return fmt.Errorf("not connected")
	}

	requestID := fmt.Sprintf("%d", time.Now().UnixNano())
	message := map[string]any{
		"type":      "command",
		"requestId": requestID,
		"command":   command,
	}
	return c.conn.WriteJSON(message)
}

func (c *ditBrowseClient) focusCamera(cameraNumber int) error {
	if cameraNumber < 1 {
		return fmt.Errorf("camera number must be >= 1")
	}
	return c.sendCommand(map[string]any{
		"type":         "focusCamera",
		"cameraNumber": cameraNumber,
	})
}

func (c *ditBrowseClient) showGrid() error {
	return c.sendCommand(map[string]any{"type": "showGrid"})
}

func (c *ditBrowseClient) toggleExpansion() error {
	return c.sendCommand(map[string]any{"type": "toggleExpansion"})
}

func (c *ditBrowseClient) requestStatus() error {
	return c.sendCommand(map[string]any{"type": "status"})
}

func uniqueCameraNumbers(status ditBrowseStatus) []int {
	seen := map[int]struct{}{}
	out := make([]int, 0)
	for _, tab := range status.Tabs {
		if tab.CameraNumber == nil || *tab.CameraNumber < 1 {
			continue
		}
		n := *tab.CameraNumber
		if _, ok := seen[n]; ok {
			continue
		}
		seen[n] = struct{}{}
		out = append(out, n)
	}
	return out
}

func selectedTitle(status ditBrowseStatus) string {
	if status.SelectedTileID == nil {
		return ""
	}
	for _, tab := range status.Tabs {
		if tab.TileID == *status.SelectedTileID {
			return tab.Title
		}
	}
	return ""
}

File: companion-module-lightlab-ditbrowse/src/protocol.ts
export const CONTROL_PROTOCOL = 'ditbrowse.control'
export const CONTROL_PROTOCOL_VERSION = 1
export const CONTROL_WEBSOCKET_PATH = '/api/ws'

export interface DitBrowseStatusTab {
	index: number
	tileId: string
	cameraId: string | null
	cameraNumber: number | null
	title: string
	url: string
}

export interface DitBrowseStatus {
	expansionEnabled: boolean
	focusMode: boolean
	selectedCameraNumber: number | null
	selectedTileId: string | null
	selectedIndex: number | null
	tabs: DitBrowseStatusTab[]
}

export type DitBrowseCommand =
	| { type: 'status' }
	| { type: 'focusCamera'; cameraNumber: number }
	| { type: 'showGrid' }
	| { type: 'toggleExpansion' }

export interface ClientHello {
	type: 'hello'
	protocol: typeof CONTROL_PROTOCOL
	protocolVersion: typeof CONTROL_PROTOCOL_VERSION
	client: { name: string; version: string }
}

export interface CommandMessage {
	type: 'command'
	requestId: string
	command: DitBrowseCommand
}

export interface ServerHello {
	type: 'hello'
	protocol: typeof CONTROL_PROTOCOL
	protocolVersion: typeof CONTROL_PROTOCOL_VERSION
	server: { name: string; version: string }
	capabilities: string[]
}

export type ResultMessage =
	| { type: 'result'; requestId: string; ok: true; status?: DitBrowseStatus }
	| {
			type: 'result'
			requestId: string
			ok: false
			error: { code: string; message: string }
	  }

export interface StatusEvent {
	type: 'event'
	event: 'status'
	revision: number
	status: DitBrowseStatus
}

export interface ProtocolErrorMessage {
	type: 'error'
	error: { code: string; message: string }
}

export type ServerMessage = ServerHello | ResultMessage | StatusEvent | ProtocolErrorMessage

export class ProtocolValidationError extends Error {}

function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function isNullableString(value: unknown): value is string | null {
	return value === null || typeof value === 'string'
}

function isNullablePositiveInteger(value: unknown): value is number | null {
	return value === null || (typeof value === 'number' && Number.isSafeInteger(value) && value >= 1)
}

function parseStatusTab(value: unknown): DitBrowseStatusTab {
	if (
		!isRecord(value) ||
		typeof value.index !== 'number' ||
		!Number.isSafeInteger(value.index) ||
		value.index < 1 ||
		typeof value.tileId !== 'string' ||
		!isNullableString(value.cameraId) ||
		!isNullablePositiveInteger(value.cameraNumber) ||
		typeof value.title !== 'string' ||
		typeof value.url !== 'string'
	) {
		throw new ProtocolValidationError('Invalid DIT Browse status tab')
	}

	return {
		index: value.index,
		tileId: value.tileId,
		cameraId: value.cameraId,
		cameraNumber: value.cameraNumber,
		title: value.title,
		url: value.url,
	}
}

export function parseStatus(value: unknown): DitBrowseStatus {
	if (
		!isRecord(value) ||
		typeof value.expansionEnabled !== 'boolean' ||
		typeof value.focusMode !== 'boolean' ||
		!isNullablePositiveInteger(value.selectedCameraNumber) ||
		!isNullableString(value.selectedTileId) ||
		!isNullablePositiveInteger(value.selectedIndex) ||
		!Array.isArray(value.tabs)
	) {
		throw new ProtocolValidationError('Invalid DIT Browse status')
	}

	return {
		expansionEnabled: value.expansionEnabled,
		focusMode: value.focusMode,
		selectedCameraNumber: value.selectedCameraNumber,
		selectedTileId: value.selectedTileId,
		selectedIndex: value.selectedIndex,
		tabs: value.tabs.map(parseStatusTab),
	}
}

function requireError(value: unknown): { code: string; message: string } {
	if (!isRecord(value) || typeof value.code !== 'string' || typeof value.message !== 'string') {
		throw new ProtocolValidationError('Invalid DIT Browse protocol error')
	}
	return { code: value.code, message: value.message }
}

export function parseServerMessage(value: unknown): ServerMessage {
	if (!isRecord(value)) {
		throw new ProtocolValidationError('DIT Browse message must be an object')
	}

	if (value.type === 'hello') {
		if (
			value.protocol !== CONTROL_PROTOCOL ||
			value.protocolVersion !== CONTROL_PROTOCOL_VERSION ||
			!isRecord(value.server) ||
			typeof value.server.name !== 'string' ||
			typeof value.server.version !== 'string' ||
			!Array.isArray(value.capabilities) ||
			!value.capabilities.every((capability) => typeof capability === 'string')
		) {
			throw new ProtocolValidationError('Unsupported DIT Browse protocol hello')
		}
		return {
			type: 'hello',
			protocol: CONTROL_PROTOCOL,
			protocolVersion: CONTROL_PROTOCOL_VERSION,
			server: { name: value.server.name, version: value.server.version },
			capabilities: value.capabilities,
		}
	}

	if (value.type === 'result') {
		if (typeof value.requestId !== 'string' || !value.requestId || typeof value.ok !== 'boolean') {
			throw new ProtocolValidationError('Invalid DIT Browse command result')
		}
		if (value.ok) {
			return {
				type: 'result',
				requestId: value.requestId,
				ok: true,
				...(value.status === undefined ? {} : { status: parseStatus(value.status) }),
			}
		}
		return {
			type: 'result',
			requestId: value.requestId,
			ok: false,
			error: requireError(value.error),
		}
	}

	if (value.type === 'event' && value.event === 'status') {
		if (typeof value.revision !== 'number' || !Number.isSafeInteger(value.revision) || value.revision < 0) {
			throw new ProtocolValidationError('Invalid DIT Browse status revision')
		}
		return {
			type: 'event',
			event: 'status',
			revision: value.revision,
			status: parseStatus(value.status),
		}
	}

	if (value.type === 'error') {
		return { type: 'error', error: requireError(value.error) }
	}

	throw new ProtocolValidationError('Unsupported DIT Browse server message')
}

export function createClientHello(version: string): ClientHello {
	return {
		type: 'hello',
		protocol: CONTROL_PROTOCOL,
		protocolVersion: CONTROL_PROTOCOL_VERSION,
		client: { name: 'companion-module-lightlab-ditbrowse', version },
	}
}

export function createCommandMessage(requestId: string, command: DitBrowseCommand): CommandMessage {
	return { type: 'command', requestId, command }
}
